ant-networking/src/relay_manager.rs

```diff
@@ -239,3 +378,227 @@ impl RelayManager {
+                match self.reservation_score.entry(*relay_server) {
+                    Entry::Occupied(mut entry) => {
+                        let stat = entry.get_mut();
+
+                        let new_value = stat.succeeded.checked_add(1);
+                        if let Some(new_value) = new_value {
+                            stat.succeeded = new_value;
+                        } else {
+                            // roll over to not saturate the value. Else the success rate would keep decreasing.
+                            stat.succeeded = 1;
+                            stat.error = 0;
+                        }
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert(ReservationStat {
+                            succeeded: 1,
+                            error: 0,
+                        });
+                    }
+                }
+            }
+
+            if !connection_success {
+                // if none of the connections have been established, we can update the stats.
+                for (relay_server, connection_id, _, result) in connections.iter() {
+                    if result.is_none() {
+                        debug!("Connection {connection_id:?} from {from_peer:?} through {relay_server:?} is still pending after {elapsed:?}. This is thrown away.");
+                        continue;
+                    };
+                    debug!("Connection {connection_id:?} from {from_peer:?} through {relay_server:?} is a failure. Increasing the error count");
+                    match self.reservation_score.entry(*relay_server) {
+                        Entry::Occupied(mut entry) => {
+                            let stat = entry.get_mut();
+
+                            let new_value = stat.error.checked_add(1);
+                            if let Some(new_value) = new_value {
+                                stat.error = new_value;
+                            } else {
+                                stat.error = 1;
+                                stat.succeeded = 0;
+                            }
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(ReservationStat {
+                                succeeded: 0,
+                                error: 1,
+                            });
+                        }
+                    }
+                }
+            }
+
+            to_remove.push(*from_peer);
+        }
+
+        for from_peer in to_remove {
+            debug!("Removing {from_peer:?} from the incoming_connections_from_remote_peer");
+            self.incoming_connections_from_remote_peer
+                .remove(&from_peer);
+        }
+
+        self.log_reservation_score();
+    }
+
+    /// Clean up the stats for relay servers that we are no longer connected to.
```

Comment:
will this make us `re-connect` to a `faulty-server` and experience another round of tracking then drop ?

Reply:
Yeah, we could add a block list here.